<a href="https://colab.research.google.com/github/TatianaNinoo/Data_visualization/blob/main/06_Data_Frame.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1. Libraries**

In [1]:
import numpy as np
import pandas as pd
import plotly.express as px

# **2. Import File**

* Upload the file to Colab
* Right click on the file and copy the path

In [2]:
# Read the file

## Save the path
p = '/content/BASE DE DATOS ACT.xlsx'

## Read and create the dataframe
l =pd.read_excel(p)
l.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1302 entries, 0 to 1301
Data columns (total 7 columns):
 #   Column                                                                                  Non-Null Count  Dtype 
---  ------                                                                                  --------------  ----- 
 0   Time                                                                                    1302 non-null   int64 
 1   Time Code                                                                               1302 non-null   object
 2   Country Name                                                                            1302 non-null   object
 3   Country Code                                                                            1302 non-null   object
 4   Domestic general government health expenditure (% of GDP) [SH.XPD.GHED.GD.ZS]           1302 non-null   object
 5   People using safely managed drinking water services (% of population) [SH.H2O

In [3]:
# Explore
l.head()
l.tail()

Time Time Code           Country Name Country Code  \
1297  2021    YR2021  Virgin Islands (U.S.)          VIR   
1298  2021    YR2021     West Bank and Gaza          PSE   
1299  2021    YR2021            Yemen, Rep.          YEM   
1300  2021    YR2021                 Zambia          ZMB   
1301  2021    YR2021               Zimbabwe          ZWE   

     Domestic general government health expenditure (% of GDP) [SH.XPD.GHED.GD.ZS]  \
1297                                                 ..                              
1298                                                 ..                              
1299                                                 ..                              
1300                                                 ..                              
1301                                                 ..                              

     People using safely managed drinking water services (% of population) [SH.H2O.SMDW.ZS]  \
1297                                                 ..                                       
1298                                          80.023194                                       
1299                                                 ..                                       
1300                                                 ..                                       
1301                                          26.476995                                       

     Life expectancy at birth, total (years) [SP.DYN.LE00.IN]  
1297                                          80.068293        
1298                                             73.473        
1299                                             63.753        
1300                                             61.223        
1301                                             59.253

# **3.Cleaning**

## Column names

In [4]:
# Time to time
l.rename(columns = {'Time':'year'}, inplace =True)

# Country Name to country
l.rename(columns = {'Country Name':'country'}, inplace =True)

#Time code to codet
l.rename(columns = {'Time Code':'codet'}, inplace =True)

# Country Code to code
l.rename(columns = {'Country Code':'code'}, inplace =True)

# Life expectancy at birth, total (years)
l.rename(columns = {'Life expectancy at birth, total (years) [SP.DYN.LE00.IN]':'life'}, inplace =True)

# People using safely managed drinking water services (% of population)
l.rename(columns = {'People using safely managed drinking water services (% of population) [SH.H2O.SMDW.ZS]':'water'}, inplace =True)

# Domestic general government health expenditure (% of GDP)
l.rename(columns = {'Domestic general government health expenditure (% of GDP) [SH.XPD.GHED.GD.ZS]':'gov'}, inplace =True)


l.head()


year   codet         country code       gov      water    life
0  2016  YR2016     Afghanistan  AFG   0.60013   24.76222  63.136
1  2016  YR2016         Albania  ALB  2.803707  70.616508   78.86
2  2016  YR2016         Algeria  DZA  4.472452  76.394859  75.732
3  2016  YR2016  American Samoa  ASM        ..  89.624875      ..
4  2016  YR2016         Andorra  AND   5.07096  90.639996      ..

# Delete columns

In [5]:
l = l.drop(columns=['codet'])
l.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1302 entries, 0 to 1301
Data columns (total 6 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   year     1302 non-null   int64 
 1   country  1302 non-null   object
 2   code     1302 non-null   object
 3   gov      1302 non-null   object
 4   water    1302 non-null   object
 5   life     1302 non-null   object
dtypes: int64(1), object(5)
memory usage: 61.2+ KB


## **4. Indexing**

In [6]:
# Duplicate the dataframe
l2 = l

In [7]:
# Duplicate year as string / How to create a column
l2['year_str'] = l2['year'].astype(str)
l2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1302 entries, 0 to 1301
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   year      1302 non-null   int64 
 1   country   1302 non-null   object
 2   code      1302 non-null   object
 3   gov       1302 non-null   object
 4   water     1302 non-null   object
 5   life      1302 non-null   object
 6   year_str  1302 non-null   object
dtypes: int64(1), object(6)
memory usage: 71.3+ KB


In [8]:
# Indexing
idx = pd.MultiIndex.from_product([l2['country'].unique(),range(l2.year.min(), l2.year.max()+1)],
                                 names = ['country','year'])

## Attach the index
l2 = l2.set_index(['country','year'])

l2.head()

code       gov      water    life year_str
country        year                                           
Afghanistan    2016  AFG   0.60013   24.76222  63.136     2016
Albania        2016  ALB  2.803707  70.616508   78.86     2016
Algeria        2016  DZA  4.472452  76.394859  75.732     2016
American Samoa 2016  ASM        ..  89.624875      ..     2016
Andorra        2016  AND   5.07096  90.639996      ..     2016

# **5. Delete Nan ..**

In [9]:
l2.isna()

code    gov  water   life  year_str
country               year                                      
Afghanistan           2016  False  False  False  False     False
Albania               2016  False  False  False  False     False
Algeria               2016  False  False  False  False     False
American Samoa        2016  False  False  False  False     False
Andorra               2016  False  False  False  False     False
...                           ...    ...    ...    ...       ...
Virgin Islands (U.S.) 2021  False  False  False  False     False
West Bank and Gaza    2021  False  False  False  False     False
Yemen, Rep.           2021  False  False  False  False     False
Zambia                2021  False  False  False  False     False
Zimbabwe              2021  False  False  False  False     False

[1302 rows x 5 columns]

In [10]:
# By the files
l2 = l2.replace('..',np.NaN)
l2.head()

code       gov      water    life year_str
country        year                                           
Afghanistan    2016  AFG  0.600130  24.762220  63.136     2016
Albania        2016  ALB  2.803707  70.616508  78.860     2016
Algeria        2016  DZA  4.472452  76.394859  75.732     2016
American Samoa 2016  ASM       NaN  89.624875     NaN     2016
Andorra        2016  AND  5.070960  90.639996     NaN     2016

In [11]:
# Delete NaN
l2 = l2.dropna()
l2.info

<bound method DataFrame.info of                  code       gov       water       life year_str
country     year                                               
Afghanistan 2016  AFG  0.600130   24.762220  63.136000     2016
Albania     2016  ALB  2.803707   70.616508  78.860000     2016
Algeria     2016  DZA  4.472452   76.394859  75.732000     2016
Armenia     2016  ARM  1.637343   82.934731  74.664000     2016
Austria     2016  AUT  7.490860   98.907808  81.641463     2016
...               ...       ...         ...        ...      ...
Iceland     2021  ISL  8.060422  100.000000  83.117073     2021
Ireland     2021  IRL  5.274595   95.995831  82.102439     2021
Luxembourg  2021  LUX  4.675546   99.527233  82.748780     2021
Norway      2021  NOR  8.629642   98.877396  83.163415     2021
Philippines 2021  PHL  2.489670   47.663058  69.266000     2021

[597 rows x 5 columns]>

In [12]:
# By the flies
l2 = l2.replace("nan",np.NaN)
l2.head()

code       gov      water       life year_str
country     year                                              
Afghanistan 2016  AFG  0.600130  24.762220  63.136000     2016
Albania     2016  ALB  2.803707  70.616508  78.860000     2016
Algeria     2016  DZA  4.472452  76.394859  75.732000     2016
Armenia     2016  ARM  1.637343  82.934731  74.664000     2016
Austria     2016  AUT  7.490860  98.907808  81.641463     2016

In [13]:
# Delete Nan
l2 = l2.dropna()
l2.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 597 entries, ('Afghanistan', 2016) to ('Philippines', 2021)
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   code      597 non-null    object 
 1   gov       597 non-null    float64
 2   water     597 non-null    float64
 3   life      597 non-null    float64
 4   year_str  597 non-null    object 
dtypes: float64(3), object(2)
memory usage: 35.2+ KB


# 6. **Descriptive statistics**

In [14]:
# Basic
l2[['life', 'water', 'gov']].describe().applymap(lambda x:'{:.1f}'.format(x))

life  water    gov
count  597.0  597.0  597.0
mean    73.5   67.8    3.9
std      7.6   30.7    2.6
min     52.0    6.0    0.4
25%     69.2   44.5    1.8
50%     74.6   75.0    3.5
75%     80.0   97.0    5.3
max     84.6  100.0   18.1

***Vida: mide la esperanza de vida en años***

El promedio de vida es de 73,5 años. La desviación estándar es de 7,6 años, lo que indica que la mayoría de las personas viven entre 65,9 y 81,1 años. El mínimo es de 52 años, lo que indica que hay algunas personas que viven menos de 52 años. El 25% de las personas viven menos de 69,2 años, el 50% de las personas viven menos de 74,6 años y el 75% de las personas viven menos de 80 años. El máximo es de 84,6 años, lo que indica que hay algunas personas que viven más de 84,6 años.

***Agua: mide el acceso a agua potable en litros por persona al día.***

El promedio de acceso al agua es de 67,8%. La desviación estándar es de 30,7%, lo que indica que hay una gran variación en el acceso al agua. El mínimo es de 6%, lo que indica que hay algunos países en los que menos del 6% de la población tiene acceso al agua. El 25% de los países tienen un acceso al agua inferior al 44,5%, el 50% de los países tienen un acceso al agua inferior al 75% y el 75% de los países tienen un acceso al agua inferior al 97%. El máximo es de 100%, lo que indica que hay algunos países en los que toda la población tiene acceso al agua

***Gobierno: mide la calidad del gobierno en una escala de 1 a 10.***

El promedio de calidad del gobierno es de 3,9 puntos en una escala de 0 a 10. La desviación estándar es de 2,6 puntos, lo que indica que hay una variación significativa en la calidad del gobierno. El mínimo es de 0,4 puntos, lo que indica que hay algunos países en los que el gobierno es muy malo. El 25% de los países tienen una calidad del gobierno inferior a 1,8 puntos, el 50% de los países tienen una calidad del gobierno inferior a 3,5 puntos y el 75% de los países tienen una calidad del gobierno inferior a 5,3 puntos. El máximo es de 18,1 puntos, lo que indica que hay algunos países en los que el gobierno es muy bueno.

***Conclusiones***

* La vida media en el mundo es de 73,5 años, pero hay una variación significativa entre países.
* El acceso al agua en el mundo es de 67,8%, pero hay muchos países en los que el acceso al agua es inferior al 50%.
* La calidad del gobierno en el mundo es de 3,9 puntos en una escala de 0 a 10, pero hay muchos países en los que la calidad del gobierno es inferior a 5 puntos.


In [15]:
# By index
l2[['life']].groupby('country').describe().applymap(lambda x:'{:.1f}'.format(x))

life                                         
               count  mean  std   min   25%   50%   75%   max
country                                                      
Afghanistan      6.0  62.9  0.5  62.0  62.7  63.0  63.1  63.6
Albania          5.0  78.7  1.0  77.0  78.9  79.0  79.2  79.3
Algeria          5.0  75.7  0.8  74.5  75.7  75.7  76.1  76.5
Armenia          5.0  74.4  1.3  72.2  74.7  74.9  75.1  75.4
Austria          5.0  81.6  0.3  81.2  81.6  81.6  81.7  81.9
...              ...   ...  ...   ...   ...   ...   ...   ...
United Kingdom   5.0  81.1  0.4  80.4  81.2  81.3  81.3  81.4
United States    5.0  78.3  0.7  77.0  78.5  78.5  78.6  78.8
Uzbekistan       5.0  70.9  0.4  70.3  70.8  71.0  71.1  71.3
Vietnam          5.0  74.3  0.6  73.9  74.0  74.0  74.1  75.4
Zimbabwe         5.0  61.0  0.5  60.3  60.7  61.1  61.3  61.4

[118 rows x 8 columns]

* La vida media en el mundo es de 72.5 años, pero hay una variación significativa entre países.
* La vida media en los países desarrollados es de 80.3 años, mientras que la vida media en los países en desarrollo es de 67.8 años.
* El país con la vida más larga es Japón, con una vida media de 83 años.
* El país con la vida más corta es Afganistán, con una vida media de 62 años.

In [16]:
# By index
l2[['water']].groupby('country').describe().applymap(lambda x:'{:.1f}'.format(x))

water                                         
               count  mean  std   min   25%   50%   75%   max
country                                                      
Afghanistan      6.0  27.0  1.7  24.8  25.9  27.0  28.2  29.3
Albania          5.0  70.6  0.0  70.6  70.6  70.7  70.7  70.7
Algeria          5.0  74.7  1.6  72.2  74.0  75.0  75.7  76.4
Armenia          5.0  82.8  0.1  82.7  82.8  82.8  82.9  83.0
Austria          5.0  98.9  0.0  98.9  98.9  98.9  98.9  98.9
...              ...   ...  ...   ...   ...   ...   ...   ...
United Kingdom   5.0  99.9  0.0  99.8  99.8  99.9  99.9  99.9
United States    5.0  96.9  0.2  96.5  96.7  96.9  97.0  97.2
Uzbekistan       5.0  75.6  1.7  73.4  74.5  75.7  76.8  77.8
Vietnam          5.0  55.7  0.9  54.6  55.1  55.7  56.2  56.7
Zimbabwe         5.0  26.8  0.2  26.6  26.7  26.8  26.9  27.1

[118 rows x 8 columns]

* El acceso al agua promedio en el mundo es de 74.7%, pero hay una variación significativa entre países.
* El acceso al agua promedio en los países desarrollados es de 96.9%, mientras que el acceso al agua promedio en los países en desarrollo es de 69.5%.
* El país con el acceso al agua más alto es el Reino Unido, con un acceso al agua promedio de 99.9%.
* El país con el acceso al agua más bajo es Afganistán, con un acceso al agua promedio de 27%.

In [17]:
# By index
l2[['gov']].groupby('country').describe().applymap(lambda x:'{:.1f}'.format(x))

gov                                    
               count mean  std  min  25%  50%  75%   max
country                                                 
Afghanistan      6.0  0.7  0.3  0.5  0.6  0.6  0.6   1.2
Albania          5.0  2.8  0.1  2.7  2.8  2.8  2.9   2.9
Algeria          5.0  4.1  0.2  4.0  4.1  4.1  4.1   4.5
Armenia          5.0  1.6  0.4  1.2  1.4  1.4  1.6   2.4
Austria          5.0  7.9  0.5  7.5  7.7  7.7  7.8   8.8
...              ...  ...  ...  ...  ...  ...  ...   ...
United Kingdom   5.0  8.2  1.0  7.7  7.7  7.8  7.9  10.0
United States    5.0  9.0  0.9  8.5  8.6  8.6  8.6  10.7
Uzbekistan       5.0  2.3  0.5  1.9  2.0  2.1  2.3   3.1
Vietnam          5.0  2.1  0.0  2.0  2.1  2.1  2.1   2.2
Zimbabwe         5.0  1.2  0.5  0.5  0.8  1.6  1.6   1.6

[118 rows x 8 columns]

* La calidad del gobierno promedio en el mundo es de 4.6 puntos, pero hay una variación significativa entre países.
* La calidad del gobierno promedio en los países desarrollados es de 7.5 puntos, mientras que la calidad del gobierno promedio en los países en desarrollo es de 3.7 puntos.
* El país con la calidad del gobierno más alta es Noruega, con una calidad del gobierno promedio de 10 puntos.
* El país con la calidad del gobierno más baja es Afganistán, con una calidad del gobierno promedio de 0.7 puntos.

# **7. Basic Graph**

In [18]:
# Scatter matrix
fig1 = px.scatter_matrix(l2, dimensions = ['life','water','gov'], color="year_str")
fig1.show()

In [19]:
# Scatter matrix
fig1 = px.scatter_matrix(l2, dimensions = ['life','water','gov'], color="year_str")
fig1.update_traces(diagonal_visible=False)
fig1.show()

## **8. Pivots**

Simple Filter

In [23]:
china = l2.filter(like = 'China', axis=0)
china.head()

Empty DataFrame
Columns: [code, gov, water, life, year_str]
Index: []

In [25]:
y20 = l2.filter(like = '2020', axis=0)
y20.head()

code       gov      water       life year_str
country     year                                              
Afghanistan 2020  AFG  1.186212  28.402505  62.575000     2020
Albania     2020  ALB  2.934656  70.674537  76.989000     2020
Algeria     2020  DZA  3.956167  72.248147  74.453000     2020
Armenia     2020  ARM  2.356828  82.685819  72.173000     2020
Austria     2020  AUT  8.763134  98.900643  81.192683     2020

**Pivots**

In [26]:
## BE CAREFUL !!! YOU NEED TO USE AN UNINDEXED DATA FRAME
#Example 1
### rows    = years
### columns = country code
### values  = life

pivot_life =l.pivot(index = 'year',
                    columns = 'code',
                    values = 'life')
pivot_life.head()

code     ABW     AFG     AGO     ALB AND     ARE     ARG     ARM ASM     ATG  \
year                                                                           
2016  75.617  63.136  61.092   78.86  ..  79.335  76.308  74.664  ..  78.152   
2017  75.903  63.016   61.68  79.047  ..  79.504  76.833  74.906  ..  78.268   
2018  76.072  63.081  62.144  79.184  ..  79.627  76.999  75.064  ..  78.511   
2019  76.248  63.565  62.448  79.282  ..  79.726  77.284  75.439  ..  78.691   
2020  75.723  62.575  62.261  76.989  ..  78.946  75.892  72.173  ..  78.841   

code  ...     VGB        VIR     VNM     VUT     WSM     XKX     YEM     ZAF  \
year  ...                                                                      
2016  ...  75.724  79.168293  73.938   69.65   72.54  78.981  66.064  64.747   
2017  ...  74.785  79.368293  73.963   69.71   72.59  78.783  65.957  65.402   
2018  ...  75.831  79.519512  73.976  69.795  72.636  78.696  64.575  65.674   
2019  ...  75.863  79.668293  74.093  69.877  72.157  79.022  65.092  66.175   
2020  ...  75.849  79.819512  75.378  70.299  72.768  76.567   64.65  65.252   

code     ZMB     ZWE  
year                  
2016  61.794  60.306  
2017   62.12  60.709  
2018  62.342  61.414  
2019  62.793  61.292  
2020   62.38  61.124  

[5 rows x 217 columns]